### PR TITLE
Enable self messages

### PR DESCRIPTION
**Pull request by mickael9**
Messages sent from another discord client will now appear as if you sent
them from Bitlbee.
    
See https://wiki.bitlbee.org/SelfMessages for more information about
this feature.

### CODE REVIEW

**Comment by dequis** [comment]
Have you tested this? What are you trying to do? That flag does nothing for groupchat messages

**Comment by mickael9** [comment]
I tested this on channels, it allows messages sent by myself from another discord client to appear as self messages on IRC (https://wiki.bitlbee.org/SelfMessages)

**Comment by dequis** [comment]
Nevermind the flag does do something, confused it with something else.

It's not strictly correct to apply it to all messages, but I guess it doesn't hurt either.

**Comment by mickael9** [comment]
Yes, I noticed that, I'm trying to improve it now since I'm not actually sure it doesn't hurt having it always. I'd also like it to work with private messages!

**Comment by sm00th** [comment]
The [patch I had for this](https://gist.github.com/sm00th/eab69cebf3400f2491b1) was a bit more complicated. The problem I had with this is described in #7: this results in echoed messages when sending from bitlbee which I see as the main usecase.

**Comment by mickael9** [comment]
I see what this is trying to do but it seems you have a race conditon between the time you receive the message info (with its id) from the HTTP POST request and the time you receive the message from the gateway (in the debug log, the gateway response appears first).

I see there is a `nonce` field that you can set in the message (and which gets repeated in the echo), we could set this to a random string and use that instead of the message id which we can't predict.

Perhaps we could do even simpler, setting the nonce to a fixed value per connection so we can ignore all self messages that we generated ourselves without having to keep track of them.

**Comment by sm00th** [comment]
> I see what this is trying to do but it seems you have a race conditon between the time you receive the message info (with its id) from the HTTP POST request and the time you receive the message from the gateway (in the debug log, the gateway response appears first).

Yeah, that was written a while ago, things probably have changed.

> I see there is a nonce field that you can set in the message (and which gets repeated in the echo), we could set this to a random string and use that instead of the message id which we can't predict.

Oh so they did add something we can use for this, nice. Haven't looked at discord's API in a while. I think that at this point we don't even need the nonce string to be unique per-session, it can just be hardcoded as "bitblee" or something like that. It could be later changed to random ids if it is decided to use this field for it's original purpose.

**Comment by mickael9** [comment]
PR updated with some inspiration from @sm00th's patch

For some reason, self messages don't work in private chats, but I think this might be an issue on the bitlbee side

File: src/discord-handlers.c [hunk 1]
```diff
@@ -494,8 +509,8 @@ static gboolean discord_prepare_message(struct im_connection *ic,
   }
 
   if (cinfo->type == CHANNEL_PRIVATE) {
-    if (g_strcmp0(author, cinfo->to.handle.name) == 0) {
-      posted = discord_post_message(cinfo, cinfo->to.handle.name, msg);
+    if (is_self || g_strcmp0(author, cinfo->to.handle.name) == 0) {
```
Comment (sm00th) [inline src/discord-handlers.c]: I think this check is not needed anymore. It is originally here to mute selfmessages, but now one of the conditions will always be true (unless I'm missing something) so we can just drop it and always call discord_post_message()

Comment (mickael9) [reply src/discord-handlers.c]: Agreed. I fixed this, let me know if you'd rather have my squash my commits

**Pull request merged**